earthfile2llb/converter.go

```diff
@@ -1799,33 +1801,89 @@ func (c *Converter) ExpandArgs(ctx context.Context, runOpts ConvertRunOpts, word
 	})
 }
 
-func (c *Converter) prepBuildTarget(ctx context.Context, fullTargetName string, platform platutil.Platform, allowPrivileged, passArgs bool, buildArgs []string, isDangling bool, cmdT cmdType) (domain.Target, ConvertOpt, bool, error) {
+func (c *Converter) absolutizeTarget(fullTargetName string, allowPrivileged bool) (domain.Target, domain.Target, bool, error) {
 	relTarget, err := domain.ParseTarget(fullTargetName)
 	if err != nil {
-		return domain.Target{}, ConvertOpt{}, false, errors.Wrapf(err, "earthly target parse %s", fullTargetName)
+		return domain.Target{}, domain.Target{}, false, errors.Wrapf(err, "earthly target parse %s", fullTargetName)
 	}
+
 	derefedTarget, allowPrivilegedImport, isImport, err := c.varCollection.Imports().Deref(relTarget)
 	if err != nil {
-		return domain.Target{}, ConvertOpt{}, false, err
+		return domain.Target{}, domain.Target{}, false, err
 	}
+
 	if isImport {
 		allowPrivileged = allowPrivileged && allowPrivilegedImport
 	}
+
 	targetRef, err := c.joinRefs(derefedTarget)
 	if err != nil {
-		return domain.Target{}, ConvertOpt{}, false, errors.Wrap(err, "join targets")
+		return domain.Target{}, domain.Target{}, false, errors.Wrap(err, "join targets")
 	}
-	target := targetRef.(domain.Target)
 
-	var pncvf variables.ProcessNonConstantVariableFunc
+	return targetRef.(domain.Target), relTarget, allowPrivileged, nil
+}
+
+func (c *Converter) checkAutoSkip(ctx context.Context, fullTargetName string, allowPrivileged, passArgs bool, buildArgs []string) (bool, func(), error) {
+
+	target, relTarget, _, err := c.absolutizeTarget(fullTargetName, allowPrivileged)
+	if err != nil {
+		return false, nil, err
+	}
+
+	overriding, _, err := c.prepOverridingVars(ctx, relTarget, passArgs, buildArgs)
+	if err != nil {
+		return false, nil, err
+	}
+
+	targetHash, _, err := inputgraph.HashTarget(ctx, inputgraph.HashOpt{
+		Target:         target,
+		Console:        c.opt.Console,
+		CI:             c.opt.IsCI,
+		BuiltinArgs:    c.opt.BuiltinArgs,
+		OverridingVars: overriding,
+	})
+	if err != nil {
+		return false, nil, errors.Wrapf(err, "auto-skip is unable to calculate hash for %s", target)
+	}
+
+	skipDB, err := bk.NewBuildkitSkipper(c.opt.AutoSkipLocalDB, c.varCollection.Org(), target, c.opt.AutoSkipClient)
```

Comment:
calling this in the converter doesn't seem great, but I understand it's needed because of the org name or target might be different.

What if instead we introduce a factory, e.g.

change `NewBuildkitSkipper` to 

```
func NewBuildkitSkipperFactory(localSkipDB, cloudClient buildkitskipper.ASKVClient) (BuildkitSkipperCreator, error) {

    return func(orgName string, target domain.Target) (BuildkitSkipper, error) {

	if localSkipDB != "" {
		skipDB, err := buildkitskipper.NewLocal(localSkipDB)
		if err != nil {
			return nil, errors.Wrapf(err, "failed to open buildkit skipper database %s", localSkipDB)
		}
		return skipDB, nil
	}

	skipDB, err := buildkitskipper.NewCloud(orgName, target, cloudClient)
	if err != nil {
		return nil, errors.Wrapf(err, "failed to create cloud-based buildkit skipper database")
	}
	return skipDB, nil
    }
}

```

it has the benefit that we don't have to propagate localSkipDB and cloudClient through to the converter.

Reply:
I think we can keep the existing structure by moving the org & target names into the methods. `NewBuildkitSkipper` is already a factory of sorts. 

As I explored this further, I realized that we can only initialize a local Bolt DB instance once. Secondary calls to `bolt.Open` block. Hence, I moved the DB initialization up & now supply it to `initAutoSkip` & as a build opt.

Reply:
> moving the org & target names into the methods

good plan!

> I realized that we can only initialize a local Bolt DB instance once

good observation!